### PR TITLE
Extract marriage-abroad payment information from templates

### PR DESCRIPTION
**Pull request by chrisroos**
Trello card: https://trello.com/c/K2bUzopZ

The Trello card is about removing payment information and links to consular fees from outcomes that don't display services/fees.

This branch removes the payment information but not the links to consular fees. I plan to deal with that in a separate branch.

Overview of the changes in this branch:

* Add full set of ceremony countries to regression test data. This allows me to use the marriage-abroad-countries-leading-to-outcome.rb script to determine the countries that might be affected by a change to the outcome template.
* Extract the payment information from the templates to the marriage_abroad_services.yml file. Update the regression test artefacts where this extraction results in changes to the outcomes.
* Introduce the _payment_for_services partial to display the relevant payment information for the ceremony country.
* Move payment partials to the payment_information_partials directory.

I would've preferred to store a set of acceptable payment methods in the YAML file (rather than the partial name) but that doesn't seem possible at the moment given how varied the payment information can be.

I've only added the payment information to the ceremony countries that already have entries in marriage_abroad_services.yml (i.e. the services/fees section is displayed for at least some of the outcomes). This ensures that we're no longer displaying the payment information on outcomes where the services/fees section isn't displayed.

While working on this branch, I used the complete set of regression test artefacts for marriage-abroad so that I could sanity check the changes I was making. I've removed these from the commits now that I'm happy with the changes that have been made.

This branch has exposed a couple of other issues.

* Outcomes for a number of ceremony countries display services/fees but no payment information.
* Outcomes for a number of ceremony countries were displaying different payment information depending on the type of ceremony. I think (hope) this is an oversight as the mechanism in this branch only allows for payment information to be specified for the country and not at a more granular level.

I've added comments to the Trello card to ask how we address both of these issues.

I plan to add another Trello card to work out whether we can reduce the number of payment information partials in the system.

### CODE REVIEW

File: lib/smart_answer/calculators/marriage_abroad_calculator.rb [hunk 1]
```diff
@@ -293,6 +293,12 @@ def services
       end
     end
 
+    def services_payment_partial_name
+      if @services_data[ceremony_country].present?
+        @services_data[ceremony_country]['payment_partial_name']
+      end
+    end
```
Comment (floehopper) [inline lib/smart_answer/calculators/marriage_abroad_calculator.rb]: I think you could use the (relatively new) Ruby [`Hash#dig`][1] method here to make the code a bit more succinct:

```ruby
def services_payment_partial_name
  @services_data.dig(ceremony_country, 'payment_partial_name')
end
```

[1]: http://ruby-doc.org/core-2.3.0_preview1/Hash.html#method-i-dig

File: lib/smart_answer_flows/marriage-abroad/outcomes/os_affirmation/_fees.govspeak.erb [hunk 2]
```diff
@@ -11,16 +11,9 @@
...
-    <%= render partial: 'pay_by_visas_or_mastercard.govspeak.erb' %>
-  <% elsif calculator.ceremony_country == 'south-korea' %>
-    <%= render partial: 'pay_by_cash_in_local_currency_or_credit_card_not_cheque' %>
-
-  <% else %>
-    <%= render partial: 'pay_by_cash_or_credit_card_no_cheque.govspeak.erb' %>
+  <% if calculator.services_payment_partial_name.present? %>
+    <% partial = calculator.services_payment_partial_name + '.govspeak.erb' %>
+    <%= render partial: partial,
+               locals: { calculator: calculator } %>
```
Comment (floehopper) [inline lib/smart_answer_flows/marriage-abroad/outcomes/os_affirmation/_fees.govspeak.erb]: Very minor: I realise you're probably trying to use the same style as other code in this template, but I think they are the relics of lines that were previously longer. I'd prefer that this call to `render` was on a single line, but I understand if you'd prefer to address it separately.

File: lib/smart_answer/calculators/marriage_abroad_calculator.rb (hunk 1 above)
Comment (floehopper) [inline lib/smart_answer/calculators/marriage_abroad_calculator.rb]: Minor: Since this is arguably presentational logic, did you consider making this a view helper method? I think I'm happy with the approach you've taken, because you've talked about ideally wanting to store the source data (as opposed to the partial template name) in the YAML.

File: lib/smart_answer_flows/marriage-abroad/outcomes/outcome_ceremonies_in_dominican_republic.govspeak.erb [hunk 3]
```diff
@@ -57,7 +57,7 @@
                locals: { calculator: calculator } %>
     <%= render partial: 'link_to_consular_fees.govspeak.erb',
                locals: { calculator: calculator } %>
-
-    <%= render partial: 'pay_by_cash_or_credit_card_no_cheque.govspeak.erb' %>
+    <%= render partial: 'payment_for_services.govspeak.erb',
+               locals: { calculator: calculator } %>
   <% end %>
 <% end %>
```
Comment (floehopper) [inline lib/smart_answer_flows/marriage-abroad/outcomes/outcome_ceremonies_in_dominican_republic.govspeak.erb]: Based on the commit note in which it was introduced, I would've expected to see changes to the `marriage-abroad-countries-leading-to-outcome.rb` script in this commit (and others like it). Am I missing something?

File: script/add-payment-partials-to-marriage-abroad-services-data.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  qatar
+  south-korea
+  thailand
+  turkey
+  united-arab-emirates
+  vietnam
+)
+
+payment_method_partial = {
+  'default'     => 'pay_by_cash_or_credit_card_no_cheque',
```
Comment (floehopper) [inline script/add-payment-partials-to-marriage-abroad-services-data.rb]: What's the purpose of this default value? It doesn't look as if it's being used in the app code.

File: lib/smart_answer_flows/marriage-abroad/outcomes/outcome_ceremonies_in_dominican_republic.govspeak.erb (hunk 3 above)
Comment (floehopper) [inline lib/smart_answer_flows/marriage-abroad/outcomes/outcome_ceremonies_in_dominican_republic.govspeak.erb]: Hmm. Having reviewed a few more commits, I now _think_ I understand why `marriage-abroad-countries-leading-to-outcome.rb` is only updated in certain commits, but it would be helpful if you could give me a quick summary.

File: test/artefacts/marriage-abroad/american-samoa/uk/partner_british/opposite_sex.txt [hunk 5]
```diff
@@ -27,8 +27,6 @@ You should also check with the local authorities in American Samoa to find out i
 
 You normally have to pay fees for consular services in the local currency - these are shown in the list of [consular fees for American Samoa](/government/publications/american-samoa-consular-fees).
 
-You can pay by cash or credit card, but not by personal cheque.
-
```
Comment (floehopper) [inline test/artefacts/marriage-abroad/american-samoa/uk/partner_british/opposite_sex.txt]: I'm not sure I understand how you've decided that changes like these to the regression test artefacts are OK.

File: lib/smart_answer_flows/marriage-abroad/outcomes/outcome_opposite_sex_marriage_in_japan.govspeak.erb [hunk 6]
```diff
@@ -55,6 +55,6 @@
              locals: { calculator: calculator } %>
 
 
-  You can pay by credit card, but not by personal cheque. The embassy in Tokyo will also accept cash.
+  <%= render partial: 'pay_by_credit_card_but_not_personal_cheque.govspeak.erb' %>
```
Comment (floehopper) [inline lib/smart_answer_flows/marriage-abroad/outcomes/outcome_opposite_sex_marriage_in_japan.govspeak.erb]: Why did you decide not to include `tokyo` in the partial name? Given that the content includes Tokyo-specific text, this seems a bit dangerous.

File: script/add-payment-partials-to-marriage-abroad-services-data.rb [hunk 7]
```diff
@@ -1,124 +1,14 @@
...
-
-  'kazakhstan' => 'pay_by_cash_in_kazakhstan',
-  'kyrgyzstan' => 'pay_by_cash_in_kazakhstan',
-
-  'luxembourg' => 'pay_in_cash_visa_or_mastercard',
-
-  'russia' => 'pay_by_mastercard_and_visa_only',
-
-  'finland' => 'pay_in_euros_or_visa_electron',
-
```
Comment (floehopper) [inline script/add-payment-partials-to-marriage-abroad-services-data.rb]: Hmm. Having thought I understood why this script changes in some commits, but not others, I was very surprised to see lines being removed in this commit. Clearly I don't understand what's going on!

**Comment by floehopper** [comment]
As you will probably have gathered from my previous comments, I found this PR a bit hard to follow:

* I may have missed it, but it would help if there was a general explanation of what the `marriage-abroad-countries-leading-to-outcome.rb` script does and how you used it.
* I think most commits in this PR are of 2 (or possibly 3) types, but all of their commit note subject lines start with the word "Extract". This makes it hard to discriminate between the different types of commit.
* The commit notes didn't seem to make it clear how you decided that the changes to the regression test artefacts were acceptable.
* I think my confusion about the changes to the regression test artefacts was made worse, because I initially overlooked the fact that the purpose of the PR is to "remove payment information from outcomes that don't display services/fees", so I was expecting this PR to just be a refactoring.
* It feels as if there is some sense in which single commits are doing *both* refactoring *and* making a change to the behaviour. Or have I got that wrong?
* In general I'm in favour of the refactoring aspect of this PR - I think it's much better if we store data about each country rather than conditionals which use lists of countries which share some random characteristic.
* It's great that you've identified some discrepancies. I hope you get some answers relatively quickly.
* I reckon it would be worthwhile flagging up some of the (accidental?) variation in the payment partials to see whether this list can be further consolidated and ideally converted into list of acceptable/unacceptable payment types for each country. The benefit to the FCO & the Content Team would be that it would be much easier to update this data.
* All in all it looks great (good work!), but I think I need to come back to it once you've explained more about the  `marriage-abroad-countries-leading-to-outcome.rb` script and the changes to the regression test artefacts. Hence I'm *not* marking it as LGTM for now.

File: lib/smart_answer/calculators/marriage_abroad_calculator.rb (hunk 1 above)
Comment (floehopper) [inline lib/smart_answer/calculators/marriage_abroad_calculator.rb]: Minor: Thinking about it a bit more, rather than use `Hash#dig`, perhaps we could extract a method like this to avoid the `if` statement:

```ruby
def ceremony_country_data
  @services_data[ceremony_country] || {}
end

def services_payment_partial_name
  ceremony_country_data['payment_partial_name']
end
```

Comment (pmanrubia) [inline lib/smart_answer/calculators/marriage_abroad_calculator.rb]: Minor: what about renaming `payment_partial_name` into `payment_type`? And getting the partial name using something like the code below in the calculator?
```ruby
def services_payment_partial_name
  ceremony_country_data['payment_type']
end
```
Not sure about talking about partials in the config file?

**Comment by chrisroos** [comment]
Thanks for reviewing, @floehopper and @pmanrubia.

I'm going to close this open a new PR that addresses your feedback.

File: lib/smart_answer/calculators/marriage_abroad_calculator.rb (hunk 1 above)
Comment (chrisroos) [inline lib/smart_answer/calculators/marriage_abroad_calculator.rb]: > Minor: Since this is arguably presentational logic, did you consider making this a view helper method? I think I'm happy with the approach you've taken, because you've talked about ideally wanting to store the source data (as opposed to the partial template name) in the YAML.

@floehopper: It didn't cross my mind to add it as a helper method. The `MarriageAbroadCalculator` felt like the natural place for it given how closely this relates to the other service information the calculator is responsible for. If we are able to store acceptable payment types, instead of names of partials, then I can imagine using the `MarriageAbroadCalculator` to retrieve the data and a new helper method to turn it into a sentence.

Comment (chrisroos) [inline lib/smart_answer/calculators/marriage_abroad_calculator.rb]: > Minor: what about renaming `payment_partial_name` into `payment_type`? 
> <snipped>
> Not sure about talking about partials in the config file?

@pmanrubia: I agree that mentioning partials in the data isn't ideal. Given that I can't see an easy way round it at the moment I prefer to make it explicit that the data represents the name of a partial to render. Using something like `payment_type` might lead someone to think they can store an arbitrary string against that key.

Comment (chrisroos) [inline lib/smart_answer/calculators/marriage_abroad_calculator.rb]: > Minor: Thinking about it a bit more, rather than use `Hash#dig`, perhaps we could extract a method like this to avoid the if statement:

@floehopper: I've extracted such a method in https://github.com/alphagov/smart-answers/pull/2507/commits/411edd96d148365e748a32ddcdead160a4d8ad16.

File: lib/smart_answer_flows/marriage-abroad/outcomes/os_affirmation/_fees.govspeak.erb (hunk 2 above)
Comment (chrisroos) [inline lib/smart_answer_flows/marriage-abroad/outcomes/os_affirmation/_fees.govspeak.erb]: > Very minor: I realise you're probably trying to use the same style as other code in this template, but I think they are the relics of lines that were previously longer. I'd prefer that this call to `render` was on a single line, but I understand if you'd prefer to address it separately.

@floehopper: I've addressed this in 
https://github.com/alphagov/smart-answers/pull/2507/commits/87263ff98791d8874164d2554cf4d158f1023d88.

File: lib/smart_answer_flows/marriage-abroad/outcomes/outcome_ceremonies_in_dominican_republic.govspeak.erb (hunk 3 above)
Comment (chrisroos) [inline lib/smart_answer_flows/marriage-abroad/outcomes/outcome_ceremonies_in_dominican_republic.govspeak.erb]: @floehopper:

The `marriage-abroad-countries-leading-to-outcome.rb` script allows me to see a list of countries that lead a specific outcome. That hasn't (or at least shouldn't have) changed in any of these commits.

The `add-payment-partials-to-marriage-abroad-services-data.rb` script was changed in a number of these commits. I used this temporary script to update marriage_abroad_services.yml data in the commits that dealt with outcomes that could be reached by more than country. In the reworked version of this branch (PR #2507), I've consistently updated and used the `add-payment-partials-to-marriage-abroad-services-data.rb` script in each commit that amends the marriage_abroad_services.yml data. Hopefully that makes its purpose and use a bit clearer!

File: script/add-payment-partials-to-marriage-abroad-services-data.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+  'philippines' => 'pay_by_cash_or_bank_cheque',
+  'south-korea' => 'pay_by_cash_in_local_currency_or_credit_card_not_cheque'
+}
+
+marriage_abroad_services_file = Rails.root.join('lib', 'data', 'marriage_abroad_services.yml')
+marriage_abroad_services = YAML.load_file(marriage_abroad_services_file)
+
+countries.each do |country|
+  payment_partial_name = payment_method_partial[country] ||
+    payment_method_partial['default']
```
Comment (chrisroos) [inline script/add-payment-partials-to-marriage-abroad-services-data.rb]: > What's the purpose of this default value? It doesn't look as if it's being used in the app code.

@floehopper: It's used on this line in the temporary script. It's the default value stored in the marriage_abroad_services.yml file when no country specific value exists.

File: test/artefacts/marriage-abroad/american-samoa/uk/partner_british/opposite_sex.txt (hunk 5 above)
Comment (chrisroos) [inline test/artefacts/marriage-abroad/american-samoa/uk/partner_british/opposite_sex.txt]: I've tried to explain why I've accepted the various regression test artefact changes in PR #2507. See https://github.com/alphagov/smart-answers/pull/2507/commits/9a91b1f78fb731d57813435ce0a6eb42c16585f2 for an explanation of these changes to American Samoa.

File: lib/smart_answer_flows/marriage-abroad/outcomes/outcome_opposite_sex_marriage_in_japan.govspeak.erb (hunk 6 above)
Comment (chrisroos) [inline lib/smart_answer_flows/marriage-abroad/outcomes/outcome_opposite_sex_marriage_in_japan.govspeak.erb]: @floehopper: I've named this partial `pay_in_japan` in https://github.com/alphagov/smart-answers/pull/2507/commits/0a13dab61a3d389c4615e829b10558f8d13711ba.

File: script/add-payment-partials-to-marriage-abroad-services-data.rb (hunk 7 above)
Comment (chrisroos) [inline script/add-payment-partials-to-marriage-abroad-services-data.rb]: Hopefully the PR description and commit notes of PR #2507 make the purpose of this script clearer!

**Comment by chrisroos** [comment]
> * I may have missed it, but it would help if there was a general explanation of what the `marriage-abroad-countries-leading-to-outcome.rb` script does and how you used it.

@floehopper: I've attempted to explain the use of this script in the following paragraph in the description of PR #2507:

> For outcomes that are rendered for multiple countries, I used the `marriage-abroad-countries-leading-to-outcome.rb` script to get a list of those countries before updating the `add-payment-partials-to-marriage-abroad-services-data.rb` script.

I've also got a separate note to add some more information about that script. Either to the script itself, or to some docs.

**Comment by chrisroos** [comment]
> * I think most commits in this PR are of 2 (or possibly 3) types, but all of their commit note subject lines start with the word "Extract". This makes it hard to discriminate between the different types of commit.

@floehopper: I've attempted to improve this slightly in PR #2507 by making it clearer when the commit extracts a partial, versus when it extracts payment information from a template. I can see that it's still not very clear but hopefully it's slightly better.

**Comment by chrisroos** [comment]
> * The commit notes didn't seem to make it clear how you decided that the changes to the regression test artefacts were acceptable.

@floehopper: I've attempted to improve this in PR #2507. See https://github.com/alphagov/smart-answers/pull/2507/commits/9a91b1f78fb731d57813435ce0a6eb42c16585f2 for an example.

**Comment by chrisroos** [comment]
> * I think my confusion about the changes to the regression test artefacts was made worse, because I initially overlooked the fact that the purpose of the PR is to "remove payment information from outcomes that don't display services/fees", so I was expecting this PR to just be a refactoring.

@floehopper: I've renamed the branch in PR #2507 to (hopefully) make its purpose clearer.

**Comment by chrisroos** [comment]
> * It feels as if there is some sense in which single commits are doing *both* refactoring *and* making a change to the behaviour. Or have I got that wrong?

@floehopper. That's correct in the case of some commits. I think it would've been possible to create a branch that just refactored how the payment information was displayed before then separately removing it for those countries that shouldn't display it. While possible, I think this would've been quite a lot more work as I would've had to support displaying different payment information for a single country. I believe this only exists by mistake (i.e. templates have been split causing differences between the information displayed for the same country).
